### ISSUE TITLE
What is __psynch_cvwait and why does it uses so much CPU?

### ISSUE THREAD

**Issue by scheglov** [dart-lang/sdk#35394]
Reproduction is not easy - requires code and configuration that is only on my machine.
  But maybe someone knows what it is.

![image](https://user-images.githubusercontent.com/384794/49888629-541f8700-fdf4-11e8-97a4-830eb6a26a7a.png)

**Comment by mraleph** [comment]
It's an implementation detail of `pthread_cond_wait` (which waits on the condition variable). 

I think in reality it is not using any CPU, but observatory profiler is sending signals to the waiting thread (because it does not know that waiting thread is not doing any actual work). 

/cc @rmacnak-google

**Comment by rmacnak-google** [comment]
In that profile, slightly less time is spent in the wait than old space GC. Old space GC does most of its work on other threads with the main thread waiting on the helper threads (for the non-concurrent parts). This should have been visible in the profile as GCOldSpace appearing as a caller of __psynch_cvwait. GCOldSpace isn't a function but rather a "tag" serving as a marker, so I suspect the table view of the profile doesn't include tags as a callees or caller. If you look at this same profile in the tree view, you should see __psynch_cvwait as a child of GCOldSpace.

This profile is worrying because GCOldSpace should only appear significant in the profile if most marking isn't running concurrently. What are you running?

**Comment by scheglov** [comment]
I don't see GCOldSpace in the tree view at all.
![image](https://user-images.githubusercontent.com/384794/49892635-51c22a80-fdfe-11e8-8c61-3a87702572ad.png)

I'm running a tiny experimental app that uses `package:watcher` to wait for changes in two folders, and then does some sync work (parse, build library scopes, walk libraries graph, etc; about 600 ms work), and some more async work (resolve affected libraries; about 150 ms work).

The command line to run:
`/Users/scheglov/Applications/dart-sdk/bin/dart --observe:5000 --code_comments=false --source_lines=false --pause_isolates_on_unhandled_exceptions=false --pause_isolates_on_exit=false --enable-asserts /Users/scheglov/Source/Dart/sdk.git/sdk/pkg/analyzer/tool/analysis_driver/incremental_dependencies.dart`, just in case if it triggers some unexpected VM flag.

![image](https://user-images.githubusercontent.com/384794/49893051-6226d500-fdff-11e8-95f6-f30ac42450c7.png)

**Comment by mraleph** [comment]
@rmacnak-google I have a related question, why don't we use `setitimer()` with `ITIMER_PROF` to implement observatory profiler? this would remove this sort of confusing profiles - because `ITIMER_PROF` would not fire when thread is sitting inside a `__psync_cvwait`.

**Comment by rmacnak-google** [comment]
@scheglov Can you try with https://dart-review.googlesource.com/c/sdk/+/87380 patched in, and say if you find it more understandable?

**Comment by scheglov** [comment]
@rmacnak-google It is not attributed to GCOldSpace. I guess, this is an improvement, although I still don't understand why it does so much old GC. Are there any usual reason why this happens? Too large heap? Too many objects? Too often new GC?

![image](https://user-images.githubusercontent.com/384794/50028297-0fd2e900-ffa4-11e8-9930-70398556d44e.png)

Here is also the allocation profile.
![image](https://user-images.githubusercontent.com/384794/50028334-355ff280-ffa4-11e8-96bb-6e9452ecd0fd.png)

**Comment by rmacnak-google** [comment]
It looks like your program is allocating a lot of long-lived lists and maps, based on large old accumulated values for _List, _GrowableList and _Uint32List (_Uint32Lists are involved in the default map implementation). The time in OldSpaceGC should go down dramatically after https://dart-review.googlesource.com/c/sdk/+/87121.

To see where these lists and maps are coming from, you could start your program paused, navigate to the _GrowableList and _InternalLinkedHashMap classes and click 'enable' under trace allocations, resume the program, then visit the classes again and click 'refresh allocation profile'.

Your program is also allocating a lot of short-lived Uint8Lists (large new accumulated value but low old accumulated value), possibly these are buffers from I/O. Short-lived objects are handled very quickly by the GC, so this is probably fine unless you don't expect the program to involve many byte buffers.

**Comment by a-siva** [comment]
is there any more actionable item in this issue? If not I would close the issue.

**Comment by scheglov** [comment]
Well, I cannot decide whether to close this issue or not.
On one hand, my original question was solved, and I have some idea why this symbol is referenced.
On the other hand, I'm still perplexed why there is so much GCOldSpace.

I'm using `Dart VM version: 2.2.0-edge.737209b85c340067898996ae88039a72f5f646a5 (Mon Dec 31 16:23:28 2018 +0000) on "macos_x64"`, and I see that https://dart-review.googlesource.com/c/sdk/+/87121 has been landed. I was told that it will bring down time in OldSpaceGC dramatically. But I still see a lot.

![image](https://user-images.githubusercontent.com/384794/50570500-fe7b3200-0d42-11e9-8b93-7cf9279637f4.png)

It seems that during the operation that I run old GC takes about `350ms`, which seems a lot for about `350MB` heap. Here are 3 states of the allocation profile screen - before the operation, after the first, and after the the same second operation. The whole operation is about `950ms`.

Is old GC time a function of the number of old objects, so no much of the heap size? Even so, `350ms` for `350MB` is seems too much. Or not? There might be 3-4 million objects in the heap.

![image](https://user-images.githubusercontent.com/384794/50570579-8d894980-0d45-11e9-9ca0-6a63100f12a1.png)

![image](https://user-images.githubusercontent.com/384794/50570583-bdd0e800-0d45-11e9-84ae-25d2d95eabe7.png)

![image](https://user-images.githubusercontent.com/384794/50570584-c88b7d00-0d45-11e9-892c-5f47b49edbfc.png)



My understanding is that the problem is not with allocating objects, but with keeping these objects for long time. Is this right? So, I don't need allocation profile, I need "reference profile". Unfortunately "heap snapshot" is not very helpful. The fact that `_List` is often references from `_GrowableList` does not tell my anything. Because, OK, but from where all these `_List` instances are referenced? And their incoming references, etc?

![image](https://user-images.githubusercontent.com/384794/50570653-cf1af400-0d47-11e9-9932-1280ca7d0296.png)

What I always would like to have in Observatory is something like https://www.yourkit.com/docs/java/help/merged_paths.jsp

**Comment by rmacnak-google** [comment]
Based on that description, I think 'merged paths' is the same thing as Observatory's 'dominator tree (merged siblings by class)'.

In any case, new heap analysis tools are a separate issue.

**Comment by scheglov** [comment]
@rmacnak-google I agree, that new heap analysis tools is a separate issue.

But there are a several of questions about old GC performance.

1. Is it expected to spend `350ms` on the heap `350MB` and `4.2 million` objects?

2. Is old GC time a function of the number of objects, and not the heap size per se? The number of references?

3. Is it expected that old GC uses so much time even after https://dart-review.googlesource.com/c/sdk/+/87121? I really don't know, just referring to the fact that it was advertised to make things better. Or is this already a better state, that I just don't appreciate enough? :-)

It is OK, if you don't want or cannot answer on these questions, just currently it seems like the issue was closed on a formal reason that something requiring a new issue, and not directly related for this issue was mentioned.

**Comment by scheglov** [comment]
Feel free to close it if you think that it would require more time to answer these questions than an issue assumes.

**Comment by mraleph** [comment]
@scheglov could you confirm the reproduction? it does not seem that `pkg/analyzer/tool/analysis_driver/incremental_dependencies.dart` is available in a normal SDK checkout, is it part of some special branch?

**Comment by scheglov** [comment]
I'm sorry, I see that I have mentioned that reproduction is not easy in the beginning, but of course we all forgot about this after the long holidays / vacations. I will prepare CL with these files and report back in this issue. Just a bit busy right now.

**Comment by scheglov** [comment]
https://dart-review.googlesource.com/c/sdk/+/88228

1. There is a hardcoded path to SDK in `pkg/analyzer/tool/analysis_driver/incremental_dependencies.dart`, change it.

2. Run it, e.g. `dart --observe:5000 --code_comments=false --source_lines=false --pause_isolates_on_unhandled_exceptions=false --pause_isolates_on_exit=false --enable-asserts /Users/scheglov/Source/Dart/sdk.git/sdk/pkg/analyzer/tool/analysis_driver/incremental_dependencies.dart`

3. Wait for analysis to finish.

4. Open `canonical_name.dart` from `kernel`. Change `CanonicalName.getChild` to `getChild2`, save and observe processing changes, re-analysis, etc. Change back to `getChild`, save.

5. Open Observatory at `127.0.0.1:5000`, go to `cpu profile (table)`,  clear statistics.

6. Do (4) several times. Refresh `cpu profile (table)` statistics. Notice that `GCOldSpace` is about `30%`.

I realize that this code generates a lot of (new?) garbage. It is just surprising how expensive `GCOldSpace` is relative to code that does change processing.

**Comment by rmacnak-google** [comment]
Oh, I missed the profile clearing. I see in Observatory's timeline that the VM is filling the pauses between processing the file changes with idle time GCs. In an idle GC, both phases are not concurrent, which explains the 2x. I imagine for your purposes it would be better if the profiler did not tick during idle GC because you're interest in where the Dart code is spending time in these incremental updates. For other purposes, like figuring out why a long lived Dart process handling a low rate of small messages has a high CPU time, it is better for the profiler to not leave it out. I need to think more about how to address this.

As a quick hack, idle GC can be disabled with `--idle_timeout_micros=0`.

**Comment by scheglov** [comment]
Thank you!

With `--idle_timeout_micros=0` I see `GCOldSpace` with just about `0.5%`.
So, I can see performance of things that matter.

So, probably it was all the way usability issue :-)

**Comment by mraleph** [comment]
@rmacnak-google did you happen to notice if those idle GCs are actually "improving" heap usage (e.g. considerably reducing fragmentation)? Because maybe we are churning CPU for no good reason on these idle GCs.
